### PR TITLE
Starting on phase 1 misc beacon changes

### PR DESCRIPTION
**Pull request by vbuterin**

### CODE REVIEW

File: specs/core/1_beacon_chain_misc.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
+```
+
+## Changes
+
+### Persistent committees
+
+Add to the beacon state the following fields:
+
+* `previous_persistent_committee_root: Hash`
```
Comment (djrtwo) [inline specs/core/1_beacon_chain_misc.md]: @CarlBeek What would be the best way to insert these fields into the beacon state? Can we use the `# @code_block` syntax or does that need to specify a function?

Comment (CarlBeek) [reply specs/core/1_beacon_chain_misc.md]: That should work fine here.

The `@codeblock` syntax basically cuts all code blocks before handling an input file and pastes as the very last step before returning the combined file. It is made for exactly this.

File: specs/core/1_beacon-chain-misc.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+## Classes
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+      shard: Shard
+      proof: List[Hash, PLACEHOLDER]
+      receipt: List[ShardReceiptDelta, PLACEHOLDER]
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]: What exactly is the `ShardReceiptDelta` object?

File: specs/core/1_beacon-chain-misc.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+<!-- /TOC -->
+
+## Classes
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+      shard: Shard
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    shard: Shard
```

Something is weird with the indentation here

File: specs/core/1_beacon-chain-misc.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+def compute_historical_state_generalized_index(earlier: ShardSlot, later: ShardSlot) -> GeneralizedIndex:
+    """
+    Computes the generalized index of the state root of slot `frm` based on the state root of slot `to`.
+    Relies on the `history_acc` in the `ShardState`, where `history_acc[i]` maintains the most recent 2**i'th
+    slot state. Works by tracing a `log(later-earlier)` step path from `later` to `earlier` through intermediate
+    blocks at the next available multiples of descending powers of two.
+    """
+    o = GeneralizedIndex(1)
+    for i in range(63, -1, -1):
+          if (later-1) & 2**i > (earlier-1) & 2**i:
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
        if (later - 1) & 2**i > (earlier - 1) & 2**i:
```

Indentation and whitespace fixes.

File: specs/core/1_beacon-chain-misc.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+              later = later - ((later-1) % 2**i) - 1
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
            later = later - ((later - 1) % 2**i) - 1
```

More indentation fixes

File: specs/core/1_beacon-chain-misc.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+              o = concat_generalized_indices(o, get_generalized_index(ShardState, 'history_acc', i))
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]: o = concat_generalized_indices(o, get_generalized_index(ShardState, 'history_acc', i))

File: specs/core/1_beacon-chain-misc.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+```python
+def process_shard_receipt(state: BeaconState, receipt_proof: ShardReceiptProof):
+    """
+    Processes a ShardReceipt object.
+    """
+    receipt_slot = state.next_shard_receipt_period[receipt_proof.shard] * SLOTS_PER_EPOCH * EPOCHS_PER_SHARD_PERIOD
+    first_slot_in_last_crosslink = state.current_crosslinks[receipt_proof.shard].start_epoch * SLOTS_PER_EPOCH
+    gindex = concat_generalized_indices(
+        get_generalized_index_of_crosslink_header(0),
+        get_generalized_index(ShardBlockHeader, 'state_root')
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
        get_generalized_index(ShardBlockHeader, 'state_root'),
```

Missing comma

File: specs/core/1_beacon-chain-misc.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        compute_historical_state_generalized_index(receipt_slot, first_slot_in_last_crosslink)
```
Comment (CarlBeek) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
        compute_historical_state_generalized_index(receipt_slot, first_slot_in_last_crosslink),
```

Missing comma

File: specs/core/1_beacon-chain-misc.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+<!-- /TOC -->
+
+## Classes
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+      shard: Shard
+      proof: List[Hash, PLACEHOLDER]
```
Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    proof: List[Hash, PLACEHOLDER]
```

File: specs/core/1_beacon-chain-misc.md (hunk 2 above)
Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    receipt: List[ShardReceiptDelta, PLACEHOLDER]
```

Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: It's defined in #1326.

File: specs/core/1_beacon-chain-misc.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    proof: List[Hash, PLACEHOLDER]
+    receipt: List[ShardReceiptDelta, PLACEHOLDER]
+```
+
+## Helpers
+
+#### `pack_compact_validator`
+
+```python
+def pack_compact_validator(index: uint64, slashed: bool, balance_in_increments: uint64) -> uint64:
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
def pack_compact_validator(index: int, slashed: bool, balance_in_increments: int) -> int:
```

File: specs/core/1_beacon-chain-misc.md [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
+    the unpacking function.
+    """
+    return (index << 16) + (slashed << 15) + balance_in_increments
+```
+
+#### `unpack_compact_validator`
+
+```python
+def unpack_compact_validator(compact_validator: uint64) -> Tuple[uint64, bool, uint64]:
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
def unpack_compact_validator(compact_validator: int) -> Tuple[int, bool, int]:
```

File: specs/core/1_beacon-chain-misc.md [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+```python
+def get_previous_power_of_2(x: int) -> int:
+    return x if x <= 2 else 2 * get_previous_power_of_2(x // 2)
+```
+
+#### `verify_merkle_proof`
+
+```python
+def verify_merkle_proof(leaf: Hash, proof: Sequence[Hash], index: GeneralizedIndex, root: Hash) -> bool:
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]: Undefined `GeneralizedIndex`.

File: specs/core/1_beacon-chain-misc.md [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    assert len(proof) == get_generalized_index_length(index)
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]: Undefined `get_generalized_index_length`

File: specs/core/1_beacon-chain-misc.md (hunk 12 above)
Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: Should be a wrapper type for int. Should not be an SSZ type, as it's never going to get serialized.

File: specs/core/1_beacon-chain-misc.md (hunk 13 above)
Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: Generalized index-related methods are defined here: https://github.com/ethereum/eth2.0-specs/pull/1186/files#diff-fb626d711756bc06ef07ecd51b009215R151

So #1186 is a dependency of this PR.

File: specs/core/1_beacon-chain-misc.md [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+### Persistent committees
+
+Add to the beacon state the following fields:
+
+```python
+# begin insert @persistent_committee_fields
+    previous_persistent_committee_root: Hash
+    current_persistent_committee_root: Hash
+    next_persistent_committee_root: Hash
```
Comment (JustinDrake) [inline specs/core/1_beacon-chain-misc.md]: Why do we need three persistent committee roots (as opposed to just one)?

File: specs/core/1_beacon-chain-misc.md [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        state.next_persistent_committee_root = hash_tree_root(committees)
+```
+
+### Shard receipt processing
+
+Add the `shard_receipts` operation to `BeaconBlockBody`:
+
+```python
+# begin insert @shard_receipts
+    shard_receipts: List[ShardReceipt, MAX_SHARD_RECEIPTS]
```
Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]: `ShardReceipt` is undefined in this spec and the shard data chains spec. It seems that the type is  implicitly `List[ShardReceiptDelta, PLACEHOLDER]`

Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]: `MAX_SHARD_RECEIPTS` is undefined. Need a `Configuration` section in this file

File: specs/core/1_beacon-chain-misc.md [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+# end insert @shard_receipts
+```
+
+Use `process_shard_receipt` to process each receipt.
+
+```python
+# begin insert @process_shard_receipts
+        (body.shard_receipts, process_shard_receipts),
```
Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]: `process_shard_receipt` takes a `ShardReceiptProof`. Should we instead define the operation added to the `BeaconState` as `shard_receipt_proofs`?

File: specs/core/1_beacon-chain-misc.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+### Persistent committees
+
+Add to the beacon state the following fields:
+
+```python
+# begin insert @persistent_committee_fields
+    previous_persistent_committee_root: Hash
+    current_persistent_committee_root: Hash
+    next_persistent_committee_root: Hash
+    next_shard_receipt_period: Vector[uint, SHARD_COUNT]
```
Comment (djrtwo) [inline specs/core/1_beacon-chain-misc.md]: Should be `uint64`
```suggestion
    next_shard_receipt_period: Vector[uint64, SHARD_COUNT]
```

File: specs/core/1_beacon-chain-misc.md (hunk 14 above)
Comment (djrtwo) [reply specs/core/1_beacon-chain-misc.md]: As discussed, let's turn this into a historical vector that updates at each period

File: specs/core/1_beacon-chain-misc.md (hunk 15 above)
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    shard_receipts: List[ShardReceiptProof, MAX_SHARD_RECEIPTS]
```

File: specs/core/1_beacon-chain-misc.md [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        state.next_persistent_committee_root = hash_tree_root(committees)
+```
+
+### Shard receipt processing
+
+Add the `shard_receipts` operation to `BeaconBlockBody`:
+
+```python
+# begin insert @shard_receipts
+    shard_receipts: List[ShardReceiptProof, MAX_SHARD_RECEIPTS]
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    shard_receipt_proofs: List[ShardReceiptProof, MAX_SHARD_RECEIPTS]
```

File: specs/core/1_beacon-chain-misc.md [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+        committees = Vector[CompactCommittee, SHARD_COUNT]([
+            committee_to_compact_committee(state, get_period_committee(state, get_current_epoch(state) + 1, i))
+            for i in range(SHARD_COUNT)
+        ])
+        state.next_persistent_committee_root = hash_tree_root(committees)
+```
+
+### Shard receipt processing
+
+Add the `shard_receipts` operation to `BeaconBlockBody`:
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
Add the `shard_receipt_proofs` operation to `BeaconBlockBody`:
```

File: specs/core/1_beacon-chain-misc.md [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+```python
+# begin insert @shard_receipts
+    shard_receipts: List[ShardReceiptProof, MAX_SHARD_RECEIPTS]
+# end insert @shard_receipts
+```
+
+Use `process_shard_receipt` to process each receipt.
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
Use `process_shard_receipt_proof` to process each receipt.
```

File: specs/core/1_beacon-chain-misc.md [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+
+```python
+# begin insert @process_shard_receipts
+        (body.shard_receipts, process_shard_receipts),
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
        (body.shard_receipt_proofs, process_shard_receipt_proofs),
```

File: specs/core/1_beacon-chain-misc.md [hunk 22]
```diff
@@ -6,43 +6,37 @@
...
-| - | - | - |
-| `GeneralizedIndex` | `uint64` | a generalized index into an SSZ merkle tree |
-
 ## Configuration
 
 | Name | Value | Unit | Duration
 | - | - | - | - |
 | `MAX_SHARD_RECEIPT_PROOFS` | `2**0` (= 1) | - | - |
 | `PERSISTENT_COMMITTEE_ROOT_LENGTH` | `2**8` (= 256) | periods | ~9 months |
+| `MICRO_REWARD` | `Gwei(2**0)` (=1) | Gwei | - |
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]: I added a stub number here. @vbuterin could you give a more precise number? Or we should use `PLACEHOLDER` here?

File: specs/core/1_beacon-chain-misc.md [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+
+<!-- /TOC -->
+
+## Configuration
+
+| Name | Value | Unit | Duration
+| - | - | - | - |
+| `MAX_SHARD_RECEIPT_PROOFS` | `2**0` (= 1) | - | - |
+| `PERSISTENT_COMMITTEE_ROOT_LENGTH` | `2**8` (= 256) | periods | ~9 months |
+| `MICRO_REWARD` | `Gwei(2**0)` (=1) | Gwei | - |
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]: I added a stub number here. @vbuterin could you give a more precise number? Or we should use `PLACEHOLDER` here?

File: specs/core/1_beacon-chain-misc.md [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+    )
+    for delta in receipt_proof.receipt:
+        if get_current_epoch(state) < state.validators[delta.index].withdrawable_epoch:
+            increase_amount = (
+                state.validators[delta.index].effective_balance * delta.reward_coefficient // REWARD_COEFFICIENT_BASE
+            )
+            increase_balance(state, delta.index, increase_amount)
+            decrease_balance(state, delta.index, delta.block_fee)
+    state.next_shard_receipt_period[receipt_proof.shard] += 1
+    increase_balance(state, get_beacon_proposer_index(state), MICRO_REWARD)
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    proposer_index = get_beacon_proposer_index(state)
    increase_balance(state, proposer_index, Gwei(get_base_reward(state, proposer_index) // MINOR_REWARD_QUOTIENT))
```

File: specs/core/1_beacon-chain-misc.md [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+
+<!-- /TOC -->
+
+## Configuration
+
+| Name | Value | Unit | Duration
+| - | - | - | - |
+| `MAX_SHARD_RECEIPT_PROOFS` | `2**0` (= 1) | - | - |
+| `PERSISTENT_COMMITTEE_ROOT_LENGTH` | `2**8` (= 256) | periods | ~9 months |
+| `MICRO_REWARD` | `Gwei(2**0)` (=1) | Gwei | - |
```
Comment (vbuterin) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
| `MINOR_REWARD_QUOTIENT` | `2**8` (=256) | - | - |
```

File: specs/core/1_beacon-chain-misc.md (hunk 23 above)
Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: See suggestions; I basically set it to be the same as the reward for including chunk challenge responses.

File: specs/core/0_beacon-chain.md [hunk 26]
```diff
@@ -1549,6 +1550,7 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
         (body.deposits, process_deposit),
         (body.voluntary_exits, process_voluntary_exit),
         (body.transfers, process_transfer),
+        # @process_shard_receipts
```
Comment (vbuterin) [inline specs/core/0_beacon-chain.md]:
```suggestion
        # @process_shard_receipt_proofs
```

File: specs/core/0_beacon-chain.md [hunk 27]
```diff
@@ -1196,6 +1196,7 @@ def process_epoch(state: BeaconState) -> None:
     # @process_reveal_deadlines
     # @process_challenge_deadlines
     process_slashings(state)
+    # @update_persistent_committee
```
Comment (vbuterin) [inline specs/core/0_beacon-chain.md]:
```suggestion
    # @update_period_committee
```

File: specs/core/1_beacon-chain-misc.md [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+```python
+def compute_historical_state_generalized_index(earlier: ShardSlot, later: ShardSlot) -> GeneralizedIndex:
+    """
+    Computes the generalized index of the state root of slot `frm` based on the state root of slot `to`.
+    Relies on the `history_acc` in the `ShardState`, where `history_acc[i]` maintains the most recent 2**i'th
+    slot state. Works by tracing a `log(later-earlier)` step path from `later` to `earlier` through intermediate
+    blocks at the next available multiples of descending powers of two.
+    """
+    o = GeneralizedIndex(1)
+    for i in range(63, -1, -1):
```
Comment (adlerjohn) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    for i in range(HISTORY_ACCUMULATOR_VECTOR - 1, -1, -1):
```

File: specs/core/1_beacon-chain-misc.md [hunk 29]
```diff
@@ -0,0 +1,242 @@
...
+```
+
+#### `process_shard_receipt_proof`
+
+```python
+def process_shard_receipt_proof(state: BeaconState, receipt_proof: ShardReceiptProof) -> None:
+    """
+    Processes a ShardReceipt object.
+    """
+    receipt_slot = state.next_shard_receipt_period[receipt_proof.shard] * SLOTS_PER_EPOCH * EPOCHS_PER_SHARD_PERIOD
```
Comment (adlerjohn) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    receipt_slot = state.next_shard_receipt_period[receipt_proof.shard] * SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH * EPOCHS_PER_SHARD_PERIOD
```

File: specs/core/1_beacon-chain-misc.md [hunk 30]
```diff
@@ -0,0 +1,242 @@
...
+    first_slot_in_last_crosslink = state.current_crosslinks[receipt_proof.shard].start_epoch * SLOTS_PER_EPOCH
```
Comment (adlerjohn) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    first_slot_in_last_crosslink = state.current_crosslinks[receipt_proof.shard].start_epoch * SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH
```

File: specs/core/1_beacon-chain-misc.md [hunk 31]
```diff
@@ -0,0 +1,242 @@
...
+```
+
+#### `committee_to_compact_committee`
+
+```python
+def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
+    """
+    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
+    """
+    validators = [state.validators[i] for i in committee]
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]:
```suggestion
    assert len(committee) <= MAX_VALIDATORS_PER_COMMITTEE
    validators = [state.validators[i] for i in committee]
```

File: specs/core/1_beacon-chain-misc.md [hunk 32]
```diff
@@ -0,0 +1,242 @@
...
+| `MINOR_REWARD_QUOTIENT` | `2**8` (=256) | - | - |
+
+## Containers
+
+#### `CompactCommittee`
+
+```python
+class CompactCommittee(Container):
+    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
```
Comment (hwwhww) [inline specs/core/1_beacon-chain-misc.md]: What do you think about adding a custom type `CompactValidator: uint64`?

File: specs/core/1_beacon-chain-misc.md [hunk 33]
```diff
@@ -0,0 +1,234 @@
...
+            leaf = hash(h + leaf)
+        else:
+            leaf = hash(leaf + h)
+    return leaf == root
+```
+
+#### `compute_historical_state_generalized_index`
+
+```python
+def compute_historical_state_generalized_index(earlier: ShardSlot, later: ShardSlot) -> GeneralizedIndex:
```
Comment (villanuevawill) [inline specs/core/1_beacon-chain-misc.md]: I put up a PR with an addition to `history_acc` and how it is computed to make this work - https://github.com/ethereum/eth2.0-specs/pull/1379

Comment (villanuevawill) [reply specs/core/1_beacon-chain-misc.md]: I simulated the algorithm with the stated change, using the following code:
```python3
def compute_historic_accumulator(slot):
    slots = [None]*14
    for i in range(1, slot + 1):
        compute_historic_accumulator_per_slot(i, slots)
    print(slots)



def compute_historic_accumulator_per_slot(slot, slots):
    depth = 63
    while depth >= 0:
        if slot % 2**depth == 0:
            slots[depth] = slot
            break
        depth -= 1


def compute_historical_state_generalized_index(earlier, later):
    for i in range(63, -1, -1):
        if (later - 1) & 2**i > (earlier - 1) & 2**i:
            later = later - ((later - 1) % 2**i) - 1
            print(i)
            print(later)
            compute_historic_accumulator(later)




compute_historic_accumulator(2393943)
compute_historical_state_generalized_index(2342, 2393943);
```
Results of the simulation:
```python3
steps to solution = 17
earlier = 2342
later = 2393943

[2393943, 2393942, 2393940, 2393928, 2393936, 2393888, 2393920, 2393728, 2393856, 2393600, 2393088, 2390016, 2387968, 2383872, 2375680, 2392064, 2293760, 2228224, 2359296, 1572864, 1048576, 2097152]
i = 21
later = 2097152
[2097151, 2097150, 2097148, 2097144, 2097136, 2097120, 2097088, 2097024, 2096896, 2096640, 2096128, 2095104, 2093056, 2088960, 2080768, 2064384, 2031616, 1966080, 1835008, 1572864, 1048576, 2097152]
i = 20
later = 1048576
[1048575, 1048574, 1048572, 1048568, 1048560, 1048544, 1048512, 1048448, 1048320, 1048064, 1047552, 1046528, 1044480, 1040384, 1032192, 1015808, 983040, 917504, 786432, 524288, 1048576]
i = 19
later = 524288
[524287, 524286, 524284, 524280, 524272, 524256, 524224, 524160, 524032, 523776, 523264, 522240, 520192, 516096, 507904, 491520, 458752, 393216, 262144, 524288]
i = 18
later = 262144
[262143, 262142, 262140, 262136, 262128, 262112, 262080, 262016, 261888, 261632, 261120, 260096, 258048, 253952, 245760, 229376, 196608, 131072, 262144]
i = 17
later = 131072
[131071, 131070, 131068, 131064, 131056, 131040, 131008, 130944, 130816, 130560, 130048, 129024, 126976, 122880, 114688, 98304, 65536, 131072]
i = 16
later = 65536
[65535, 65534, 65532, 65528, 65520, 65504, 65472, 65408, 65280, 65024, 64512, 63488, 61440, 57344, 49152, 32768, 65536]
i = 15
later = 32768
[32767, 32766, 32764, 32760, 32752, 32736, 32704, 32640, 32512, 32256, 31744, 30720, 28672, 24576, 16384, 32768]
i = 14
later = 16384
[16383, 16382, 16380, 16376, 16368, 16352, 16320, 16256, 16128, 15872, 15360, 14336, 12288, 8192, 16384]
i = 13
later = 8192
[8191, 8190, 8188, 8184, 8176, 8160, 8128, 8064, 7936, 7680, 7168, 6144, 4096, 8192]
i = 12
later = 4096
[4095, 4094, 4092, 4088, 4080, 4064, 4032, 3968, 3840, 3584, 3072, 2048, 4096]
i = 10
later = 3072
[3071, 3070, 3068, 3064, 3056, 3040, 3008, 2944, 2816, 2560, 3072, 2048]
i = 9
later = 2560
[2559, 2558, 2556, 2552, 2544, 2528, 2496, 2432, 2304, 2560, 1024, 2048]
i = 7
later = 2432
[2431, 2430, 2428, 2424, 2416, 2400, 2368, 2432, 2304, 1536, 1024, 2048]
i = 6
later = 2368
[2367, 2366, 2364, 2360, 2352, 2336, 2368, 2176, 2304, 1536, 1024, 2048]
i = 4
later = 2352
[2351, 2350, 2348, 2344, 2352, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
i = 3
later = 2344
[2343, 2342, 2340, 2344, 2320, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
i = 1
later = 2342
[2341, 2342, 2340, 2328, 2320, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
```

TLDR - it looks like the solution can be slightly optimized since the first step does not pick the optimal index. In this case, it chose `i = 21` which equated to `2097152` when it could have just chosen `i = 20` which equated to `1048576` in the first step.

Also curious - what the name of this exact algorithm in this case?

Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: It's basically putting the most recent state at slot `k * 2**i` in position `i`.

File: specs/core/1_beacon-chain-misc.md (hunk 32 above)
Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: > Remove unpack_compact_validator from 1_beacon-chain-misc.md since it's not used here.

Where would it be put though? In the light client file that actually uses it? I can see how that's theoretically better in one way, but it has a big disadvantage, which is that pack and unpack are no longer beside each other. So I'd still favor putting both in the same file, to make it easier for readers to see that they are inverses.

File: specs/core/1_beacon-chain-misc.md (hunk 31 above)
Comment (vbuterin) [reply specs/core/1_beacon-chain-misc.md]: Is this needed? This function could theoretically be used for different kinds of committees; doesn't seem wise to put a maximum designed around a specific type of committee.

Comment (hwwhww) [reply specs/core/1_beacon-chain-misc.md]: It's a safety check since the size of `CompactCommittee` elements are set to maximum `MAX_VALIDATORS_PER_COMMITTEE`:
```python
class CompactCommittee(Container):
    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
```
It leads to one more implicit config validity condition of that `MAX_VALIDATORS_PER_COMMITTEE >= TARGET_PERSISTENT_COMMITTEE_SIZE`. (p.s. #407)

If we want to make this function could be used for different kinds of committees, how about refactoring this function into:
```python
def pack_compact_validators(state: BeaconState, committee: Sequence[ValidatorIndex]) -> Tuple[Sequence[BLSPubkey], Sequence[int]]:
    validators = [state.validators[i] for i in committee]
    compact_validators = [
        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
        for i, v in zip(committee, validators)
    ]
    pubkeys = [v.pubkey for v in validators]
    return pubkeys, compact_validators


def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
    assert len(committee) <= MAX_VALIDATORS_PER_COMMITTEE
    pubkeys, compact_validators = pack_compact_validators(state, committee)
    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
```

**Comment by vbuterin** [comment]
> @vbuterin Is this ready for merge?

I'd say so. My preference is generally to merge earlier for specs that are still in-progress; it makes it easier to propose further changes.

**Comment by JustinDrake** [comment]
> Justin's upcoming PR

The PR is ready for review :) https://github.com/ethereum/eth2.0-specs/pull/1383

File: specs/core/1_beacon-chain-misc.md [hunk 34]
```diff
@@ -0,0 +1,238 @@
...
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
+```
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+    shard: Shard
+    proof: List[Hash, PLACEHOLDER]
+    receipt: List[ShardReceiptDelta, PLACEHOLDER]
```
Comment (terencechain) [inline specs/core/1_beacon-chain-misc.md]: Looks like `ShardReceiptDelta` is going away in #1383. Should we define the `ShardReceiptDelta` container here?

Comment (djrtwo) [reply specs/core/1_beacon-chain-misc.md]: Going to merge this and handle the conflict after reviewing #1383. Thanks for pointing this out :)
